lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1262,7 +1262,7 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
     // check the kind of type this discard statement appears within.
     if (!diagnosed) {
       auto *nominalDecl = fn->getDeclContext()->getSelfNominalTypeDecl();
-      Type nominalType = nominalDecl->getDeclaredType();
+      Type nominalType = nominalDecl->getDeclaredTypeInContext();
 
       // must be noncopyable
```

Comment:
Why not just call isMoveOnly() on the nominal declaration?

Reply:
This line's specific change is actually to fix a subtle issue a few lines below where the `nominalType` is used to specify the `constraints::ContextualTypeInfo` in the context of a `CTP_DiscardStmt`. The type checker was complaining about failing to match `HasGeneric` with expected contextual type `HasGeneric<T>`. So I figured I need to give it the expected bound type rather than an unbound type when typechecking the expr in `discard <expr>`.

That problem was uncovered after improving the `isPureMoveOnly` check to also consider unbound generic types, since we were simply considering a noncopyable generic as copyable in this specific check.

Reply:
> This line's specific change is actually to fix a subtle issue a few lines below where the nominalType is used to specify the constraints::ContextualTypeInfo in the context of a CTP_DiscardStmt. 

Ah, in that case you shouldn't use nominal->getDeclaredTypeInContext(), because it will use the wrong generic environment (the one for the nominal) if the consuming method has its own generic parameter list or where clause (in which case the method has a different generic environment).

I think you want `fn->getImplicitSelfDecl()->getType()`, or `fn->mapTypeIntoContext(nominal->getDeclaredInterfaceType())`.

(A good rule of thumb is to just avoid getDeclaredType() and getDeclaredTypeInContext() entirely. getDeclaredInterfaceType() is your friend.)

Reply:
ok ill make that change in a follow-up PR. thanks!

Reply:
https://github.com/apple/swift/pull/65931